File: cables.py
import numpy as np
import materials as mat

class Layer:
    material: mat.Material
    diameter: float
    
    def __init__(self, material, diameter):
        self.material = material
        self.diameter = diameter

class Cable:
    cable_ref: str
    outer: Layer
    dielectric: Layer
    inner: Layer

    def __init__(self, cable_ref, outer, dielectric, inner):
        self.cable_ref = cable_ref
        self.outer = outer
        self.dielectric = dielectric
        self.inner = inner

    @property
    def outer_area(self):
        R1 = self.outer.diameter / 2
        R2 = self.dielectric.diameter / 2

        return np.pi*(R1**2 - R2**2)

    @property
    def dielectric_area(self):
        R1 = self.dielectric.diameter / 2
        R2 = self.inner.diameter / 2

        return np.pi*(R1**2 - R2**2)

    @property
    def inner_area(self):
        R1 = self.inner.diameter / 2

        return np.pi*(R1**2)

class Cables_DB:
    db: dict[str, Cable] = {}

    def __init__(self, cables: list[Cable]):
        for cable in cables:
            self.db[cable.cable_ref] = cable

Radiall_NbTi_cable = Cable(
    cable_ref = 'Radiall NbTi',
    outer=Layer(
        material=mat.materials.db['NbTi'],
        diameter=0.9e-3,
    ),
    dielectric=Layer(
        material=mat.materials.db['PTFE'],
        diameter=0.66e-3,
    ),
    inner=Layer(
        material=mat.materials.db['NbTi'],
        diameter=0.203e-3,
    ),
)

SC_119_50_SCN_CN = Cable(
    cable_ref = 'SC-119/50-SCN-CN',
    outer=Layer(
        material=mat.materials.db['SCN'],
        diameter=1.19e-3,
    ),
    dielectric=Layer(
        material=mat.materials.db['PTFE'],
        diameter=0.94e-3,
    ),
    inner=Layer(
        material=mat.materials.db['CN'],
        diameter=0.287e-3,
    ),
)

SC_86_50_SCN_CN = Cable(
    cable_ref = 'SC-86/50-SCN-CN',
    outer=Layer(
        material=mat.materials.db['SCN'],
        diameter=0.86e-3,
    ),
    dielectric=Layer(
        material=mat.materials.db['PTFE'],
        diameter=0.66e-3,
    ),
    inner=Layer(
        material=mat.materials.db['CN'],
        diameter=0.203e-3,
    ),
)

SC_86_50_NbTi_NbTi = Cable(
    cable_ref = 'SC-86/50-NbTi-NbTi',
    outer=Layer(
        material=mat.materials.db['NbTi'],
        diameter=0.90e-3,
    ),
    dielectric=Layer(
        material=mat.materials.db['PTFE'],
        diameter=0.66e-3,
    ),
    inner=Layer(
        material=mat.materials.db['NbTi'],
        diameter=0.203e-3,
    ),
)

PhBr_36AWG = Cable(
    cable_ref= 'PhBr_36AWG',
    outer = Layer(material=mat.materials.db['PhBr'], diameter = 0.127e-3,),
    dielectric = Layer(material=mat.materials.db['PTFE'], diameter = 0.127e-3,),
    inner = Layer(material=mat.materials.db['PhBr'], diameter = 0.127e-3,)
)

NbTi_36AWG = Cable(
    cable_ref= 'NbTi_36AWG',
    outer = Layer(material=mat.materials.db['NbTi'], diameter = 0.127e-3,),
    dielectric = Layer(material=mat.materials.db['NbTi'], diameter = 0.127e-3,),
    inner = Layer(material=mat.materials.db['NbTi'], diameter = 0.127e-3,)
)

cables = Cables_DB([Radiall_NbTi_cable, SC_119_50_SCN_CN, SC_86_50_SCN_CN,
                    SC_86_50_NbTi_NbTi, PhBr_36AWG, NbTi_36AWG])

File: src/cables.rs
//! Dataclass for storing the different cable types

use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::f64::consts::PI;

use crate::materials::{Material, MATERIALS_DB};

/// Layer struct
#[derive(Debug, Clone)]
pub struct Layer{
    ///Material type
    pub material: Material,
    /// Diameter
    pub diameter: f64,
}

impl Layer{
    /// Initialize new layer
    pub fn new(material:Material, diameter: f64) -> Self{
        Self { material, diameter}
    }
}

#[derive(Debug, Clone)]
/// Cable struct
pub struct Cable{
    /// Cable Reference
    pub cable_ref: String,
    /// Outer diameter
    pub outer: Layer,
    /// Dielectric type
    pub dielectric: Layer,
    /// Inner diameter
    pub inner: Layer,
}

impl Cable{
    /// Initialize new cable
    pub fn new(cable_ref:String, outer:Layer, dielectric: Layer, inner: Layer) -> Self{
        Self { cable_ref, outer, dielectric,inner }
    }
    /// Calculate outer area 
    pub fn outer_area(&self) -> f64 {
        let r1 = self.outer.diameter / 2.0;
        let r2 = self.dielectric.diameter / 2.0;
        PI * (r1 * r1 - r2 * r2)
    }
    /// Calculate dielectric area 
    pub fn dielectric_area(&self) -> f64 {
        let r1 = self.dielectric.diameter / 2.0;
        let r2 = self.inner.diameter / 2.0;
        PI * (r1 * r1 - r2 * r2)
    }
    /// Calculate inner area 
    pub fn inner_area(&self) -> f64 {
        let r = self.inner.diameter / 2.0;
        PI * r * r
    }
}

// ----- DB type -----

/// Cables database
#[derive(Debug, Clone)]
pub struct CablesDB {
    /// Dictionary
    pub db: HashMap<String, Cable>,
}

impl CablesDB {
    /// Initialization function
    pub fn new(cables: Vec<Cable>) -> Self {
        let mut db = HashMap::new();
        for cable in cables {
            db.insert(cable.cable_ref.clone(), cable);
        }
        Self { db }
    }
    /// Getter function
    pub fn get(&self, id: &str) -> Option<&Cable> {
        self.db.get(id)
    }
}

// ----- Helper to fetch & clone a Material from the materials DB -----

fn mat(id: &str) -> Material {
    MATERIALS_DB
        .get(id)
        .unwrap_or_else(|| panic!("Unknown material id: {id}"))
        .clone()
}

// ----- Prebuilt cables (global, like your Python module-level objects) -----
/// Static cable database
pub static CABLES_DB: Lazy<CablesDB> = Lazy::new(|| {
    let radiall_nbti_cable = Cable::new(
        "Radiall NbTi".to_string(),
        Layer::new(mat("NbTi"), 0.9e-3),
        Layer::new(mat("PTFE"), 0.66e-3),
        Layer::new(mat("NbTi"), 0.203e-3),
    );

    let sc_119_50_scn_cn = Cable::new(
        "SC-119/50-SCN-CN".to_string(),
        Layer::new(mat("SCN"), 1.19e-3),
        Layer::new(mat("PTFE"), 0.94e-3),
        Layer::new(mat("CN"), 0.287e-3),
    );

    let sc_86_50_scn_cn = Cable::new(
        "SC-86/50-SCN-CN".to_string(),
        Layer::new(mat("SCN"), 0.86e-3),
        Layer::new(mat("PTFE"), 0.66e-3),
        Layer::new(mat("CN"), 0.203e-3),
    );

    let sc_86_50_nbti_nbti = Cable::new(
        "SC-86/50-NbTi-NbTi".to_string(),
        Layer::new(mat("NbTi"), 0.90e-3),
        Layer::new(mat("PTFE"), 0.66e-3),
        Layer::new(mat("NbTi"), 0.203e-3),
    );

    let phbr_36awg = Cable::new(
        "PhBr_36AWG".to_string(),
        Layer::new(mat("PhBr"), 0.127e-3),
        Layer::new(mat("PTFE"), 0.127e-3),
        Layer::new(mat("PhBr"), 0.127e-3),
    );

    let nbti_36awg = Cable::new(
        "NbTi_36AWG".to_string(),
        Layer::new(mat("NbTi"), 0.127e-3),
        Layer::new(mat("NbTi"), 0.127e-3),
        Layer::new(mat("NbTi"), 0.127e-3),
    );

    CablesDB::new(vec![
        radiall_nbti_cable,
        sc_119_50_scn_cn,
        sc_86_50_scn_cn,
        sc_86_50_nbti_nbti,
        phbr_36awg,
        nbti_36awg,
    ])
});